patches/tModLoader/Terraria/ModLoader/Core/AssemblyManager.cs

```diff
@@ -334,16 +334,32 @@ internal static bool FirstModInStackTrace(StackTrace stack, out string modName)
 
 	private static IDictionary<Assembly, Type[]> GetLoadableTypes(ModLoadContext mod, MetadataLoadContext mlc)
 	{
+		Type loadingType = null;
 		try {
 			return mod.Assemblies.ToDictionary(a => a, asm =>
 				mlc.LoadFromAssemblyName(asm.GetName()).GetTypes()
 					.Where(mType => IsLoadable(mod, mType))
-					.Select(mType => asm.GetType(mType.FullName, throwOnError: true, ignoreCase: false))
+					.Select(mType => {
+						loadingType = mType;
+						return asm.GetType(mType.FullName, throwOnError: true, ignoreCase: false);
```

Comment:
Instead of storing `loadingType` outside the `Select` clause, can you extract the `asm.GetType` call to a separate method `GetType(asm, mType)` which contains the `try/catch`? You can then annotate the thrown exception in the same way that `IsLoadable` does.